pkg/webhook/parser_test.go

```diff
@@ -0,0 +1,1275 @@
+	gogsclient "github.com/gogits/go-gogs-client"
+	"gopkg.in/go-playground/webhooks.v5/bitbucket"
+	bitbucketserver "gopkg.in/go-playground/webhooks.v5/bitbucket-server"
+	"gopkg.in/go-playground/webhooks.v5/github"
+	"gopkg.in/go-playground/webhooks.v5/gitlab"
+	"gotest.tools/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+func TestParseGogs(t *testing.T) {
+	tests := map[string]struct {
+		secretData map[string][]byte
+		body       []byte
+		headers    map[string]string
+		wantErr    bool
+		wantErrMsg string
+		wantEvent  interface{}
+	}{
+		"valid-gogs-push-event-no-secret": {
+			secretData: nil,
+			body: []byte(`{
+				"ref": "refs/heads/main",
+				"after": "af69d162de5a276abc86e0686b2b44033cd3f442",
+				"repository": {
+					"html_url": "https://gogs.example.com/example/repo"
+				}
+			}`),
+			headers: map[string]string{
+				"X-Gogs-Event": "push",
+			},
+			wantErr: false,
+			wantEvent: gogsclient.PushPayload{
+				Ref:   "refs/heads/main",
+				After: "af69d162de5a276abc86e0686b2b44033cd3f442",
+				Repo: &gogsclient.Repository{
+					HTMLURL: "https://gogs.example.com/example/repo",
+				},
+			},
+		},
+		"valid-gogs-push-event-with-secret": {
+			secretData: map[string][]byte{
+				gogsKey: []byte("gogssecret"),
+			},
+			body: []byte(`{
+				"ref": "refs/heads/main",
+				"after": "af69d162de5a276abc86e0686b2b44033cd3f442",
+				"repository": {
+					"html_url": "https://gogs.example.com/example/repo"
+				}
+			}`),
+			headers: map[string]string{
+				"X-Gogs-Event":     "push",
+				"X-Gogs-Signature": "bd00de14f93a8bef25c89bb3c976a1320b3a535515cd66c91ec1e8b9f67a2259",
+			},
+			wantErr: false,
+			wantEvent: gogsclient.PushPayload{
+				Ref:   "refs/heads/main",
+				After: "af69d162de5a276abc86e0686b2b44033cd3f442",
+				Repo: &gogsclient.Repository{
+					HTMLURL: "https://gogs.example.com/example/repo",
+				},
+			},
+		},
+		"invalid-gogs-push-event-with-secret": {
+			secretData: map[string][]byte{
+				gogsKey: []byte("gogssecret"),
+			},
+			body: []byte(`{
+				"ref": "refs/heads/main",
+				"after": "af69d162de5a276abc86e0686b2b44033cd3f442",
+				"repository": {
+					"html_url": "https://gogs.example.com/example/repo"
+				}
+			}`),
+			headers: map[string]string{
+				"X-Gogs-Event":     "push",
+				"X-Gogs-Signature": "wrongsignature",
+			},
+			wantErr:    true,
+			wantErrMsg: "HMAC verification failed",
+			wantEvent:  nil,
+		},
+		"missing-gogs-secret": {
+			secretData: map[string][]byte{
+				"wrongkey": []byte("gogssecret"),
+			},
+			body: []byte(`{
+				"ref": "refs/heads/main",
+				"after": "af69d162de5a276abc86e0686b2b44033cd3f442",
+				"repository": {
+					"html_url": "https://gogs.example.com/example/repo"
+				}
+			}`),
+			headers: map[string]string{
+				"X-Gogs-Event":     "push",
+				"X-Gogs-Signature": "sha1=57968570459352679905f2372454008f9015101a",
```

Comment:
This is not very relevant to this test case, but why is the signature format different here?
Also, why `sha1`? I thought that algorithm was deprecated.

Reply:
It's simpler... :P I did copy/paste that from a Github test.
[Github still supports HMAC-SHA1 for legacy purposes](https://docs.github.com/en/webhooks/using-webhooks/validating-webhook-deliveries#troubleshooting) and the [library](https://github.com/go-playground/webhooks/blob/59fb74817c3d5c702553964f4d2faff80f23d350/github/github.go#L157:L160) we use for parsing the webhook uses the legacy header (which is HMAC-SHA1).
The latest version of the library (v6) still uses the HMAC-SHA1 header, btw.

Anyway... at this particular test the signature is irrelevant (as you noticed) because we're not even reading it, but I changed it so it does not say sha1=

Reply:
Nice, thanks for clarifying :bow: